File: frameworks/customer-storage/bucket-for-tenant.ts
// Canonical bucketForTenant — keep in sync with bucket-for-tenant.py.
// See README + ADR-004 in the WAVE control plane.

const POOL_PREFIX = "wave-customer-storage-pool-";
const DEFAULT_POOL_SIZE = 10;

function poolSize(): number {
  const env = typeof process !== "undefined" ? process.env?.WAVE_STORAGE_POOL_SIZE : undefined;
  const n = env ? Number.parseInt(env, 10) : DEFAULT_POOL_SIZE;
  return Number.isFinite(n) && n > 0 ? n : DEFAULT_POOL_SIZE;
}

export async function bucketForTenant(tenantId: string): Promise<string> {
  if (!tenantId) {
    throw new Error("bucketForTenant: tenant_id required");
  }
  const data = new TextEncoder().encode(tenantId);
  const hash = await crypto.subtle.digest("SHA-256", data);
  const index = new DataView(hash).getUint32(0, false) % poolSize();
  return `${POOL_PREFIX}${index}`;
}

export function allBucketNames(): readonly string[] {
  const n = poolSize();
  return Array.from({ length: n }, (_, i) => `${POOL_PREFIX}${i}`);
}

File: frameworks/customer-storage/bucket_for_tenant.py
"""Canonical bucket_for_tenant — keep in sync with bucket-for-tenant.ts.

See README + ADR-004 in the WAVE control plane.
"""

from __future__ import annotations

import hashlib
import os

_POOL_PREFIX = "wave-customer-storage-pool-"
_DEFAULT_POOL_SIZE = 10


def _pool_size() -> int:
    raw = os.environ.get("WAVE_STORAGE_POOL_SIZE")
    if raw is None:
        return _DEFAULT_POOL_SIZE
    try:
        n = int(raw)
        return n if n > 0 else _DEFAULT_POOL_SIZE
    except ValueError:
        return _DEFAULT_POOL_SIZE


def bucket_for_tenant(tenant_id: str) -> str:
    """Return wave-customer-storage-pool-N where N = SHA-256(tenant_id) % pool_size.

    Same as TS bucketForTenant — must produce identical output.
    """
    if not tenant_id:
        raise ValueError("bucket_for_tenant: tenant_id required")
    digest = hashlib.sha256(tenant_id.encode("utf-8")).digest()
    index = int.from_bytes(digest[:4], byteorder="big") % _pool_size()
    return f"{_POOL_PREFIX}{index}"


def all_bucket_names() -> list[str]:
    n = _pool_size()
    return [f"{_POOL_PREFIX}{i}" for i in range(n)]
